### PR TITLE
nix: Patch config.nix.in to reference host platform binaries

### PR DESCRIPTION
**Pull request by eamsden**
###### Motivation for this change
Presently the stable version of nix (2.3.3), when cross-compiled from nixpkgs, will pull into its closure and attempt to invoke the build platform's version of bash. This works around some of the build system for nix to correct that, so that a cross-compiled nix binary can be used on the system it is compiled for.

See #81153 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/package-management/nix/default.nix [hunk 1]
```diff
@@ -74,7 +75,18 @@ common =
...
+          mkdir tmp/
+          substitute corepkgs/config.nix.in tmp/config.nix.in \
+	    --subst-var-by bash ${bash}/bin/bash \
+	    --subst-var-by coreutils ${coreutils} \
+	    --subst-var-by bzip2 ${bzip2}/bin/bzip2 \
+	    --subst-var-by gzip ${gzip}/bin/gzip \
+	    --subst-var-by xz ${xz}/bin/xz \
+	    --subst-var-by tar ${gnutar}/bin/tar \
+	    --subst-var-by tr ${coreutils}/bin/tr
+	  mv tmp/config.nix.in corepkgs/config.nix.in
```
Comment (flokli) [inline pkgs/tools/package-management/nix/default.nix]: Can you indent that properly?

File: pkgs/tools/package-management/nix/default.nix [hunk 2]
```diff
@@ -74,7 +75,18 @@ common =
           ''}
         '' else ''
           configureFlagsArray+=(BDW_GC_LIBS="-lgc -lgccpp")
```
Comment (flokli) [inline pkgs/tools/package-management/nix/default.nix]: That could be part of `configureFlags` and outside `preConfigure`.

File: pkgs/tools/package-management/nix/default.nix [hunk 3]
```diff
@@ -64,7 +65,7 @@ common =
       preConfigure =
         # Copy libboost_context so we don't get all of Boost in our closure.
         # https://github.com/NixOS/nixpkgs/issues/45462
-        if is20 then ''
+        (if is20 then ''
           mkdir -p $out/lib
           cp -pd ${boost}/lib/{libboost_context*,libboost_thread*,libboost_system*} $out/lib
           rm -f $out/lib/*.a
```
Comment (flokli) [inline pkgs/tools/package-management/nix/default.nix]: you can use `optionalString` here.

File: pkgs/tools/package-management/nix/default.nix (hunk 2 above)
Comment (eamsden) [reply pkgs/tools/package-management/nix/default.nix]: This expression and the one below were already present, I added only the parens.

File: pkgs/tools/package-management/nix/default.nix (hunk 3 above)
Comment (eamsden) [reply pkgs/tools/package-management/nix/default.nix]: See above

File: pkgs/tools/package-management/nix/default.nix (hunk 2 above)
Comment (flokli) [reply pkgs/tools/package-management/nix/default.nix]: Sure, but some drive-by cleanup in a follow-up commit would be more than welcome :-)

File: pkgs/tools/package-management/nix/default.nix (hunk 1 above)
Comment (flokli) [reply pkgs/tools/package-management/nix/default.nix]: To be precise, the `--subst-var-by` lines should be 2 lines more indented than the `substitute` above, and the `mv` on the same level as `substitute` itself.

Comment (eamsden) [reply pkgs/tools/package-management/nix/default.nix]: Dammit my editor snuck tabs in

File: pkgs/tools/package-management/nix/default.nix (hunk 2 above)
Comment (eamsden) [reply pkgs/tools/package-management/nix/default.nix]: OK. I wasn't sure if that was the case. Other OSS projects have different views on such things.

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

File: pkgs/tools/package-management/nix/default.nix (hunk 2 above)
Comment (flokli) [reply pkgs/tools/package-management/nix/default.nix]: Please make that a separate commit, so we can backport only the fixes to 20.03.

File: pkgs/tools/package-management/nix/default.nix (hunk 1 above)
Comment (flokli) [reply pkgs/tools/package-management/nix/default.nix]: It seemed it did that again - can you take another look?

File: pkgs/tools/package-management/nix/default.nix [hunk 4]
```diff
@@ -64,17 +65,29 @@ common =
       preConfigure =
         # Copy libboost_context so we don't get all of Boost in our closure.
         # https://github.com/NixOS/nixpkgs/issues/45462
-        if is20 then ''
+	lib.optionalString is20 ''
```
Comment (flokli) [inline pkgs/tools/package-management/nix/default.nix]: this still doesn't look properly indented.

File: pkgs/tools/package-management/nix/default.nix [hunk 5]
```diff
@@ -64,17 +65,29 @@ common =
...
           rm -f $out/lib/*.a
           ${lib.optionalString stdenv.isLinux ''
             chmod u+w $out/lib/*.so.*
             patchelf --set-rpath $out/lib:${stdenv.cc.cc.lib}/lib $out/lib/libboost_thread.so.*
           ''}
-        '' else ''
-          configureFlagsArray+=(BDW_GC_LIBS="-lgc -lgccpp")
-        '';
+        '' +
+	# For Nix-2.3, patch around an issue where the Nix configure step pulls in the
```
Comment (flokli) [inline pkgs/tools/package-management/nix/default.nix]: same as above.

File: pkgs/tools/package-management/nix/default.nix [hunk 6]
```diff
@@ -87,6 +100,7 @@ common =
           "--with-dbi=${perlPackages.DBI}/${perl.libPrefix}"
           "--with-dbd-sqlite=${perlPackages.DBDSQLite}/${perl.libPrefix}"
           "--with-www-curl=${perlPackages.WWWCurl}/${perl.libPrefix}"
+	  "BDW_GC_LIBS=\"-lgc -lgccpp\""
```
Comment (flokli) [inline pkgs/tools/package-management/nix/default.nix]: same as above.

**Review by flokli — CHANGES_REQUESTED** [review]
Can you also split the patching of config.nix.in and the cleanup into two separate commits, so we can only backport the fix into 20.03?

File: pkgs/tools/package-management/nix/default.nix (hunk 4 above)
Comment (eamsden) [reply pkgs/tools/package-management/nix/default.nix]: I finally got neovim to not indent with tabs just this morning. Sorry. Ran `grep -Pn '\t' on the file in question to make sure no more slipped through

File: pkgs/tools/package-management/nix/default.nix (hunk 5 above)
Comment (eamsden) [reply pkgs/tools/package-management/nix/default.nix]: Done

**Comment by eamsden** [comment]
@flokli I split into two commits and made sure no more tabs got in. Can you re-review?

**Comment by flokli** [comment]
I built this, and `nix-store -qR $(nix-build -A pkgsCross.armv7l-hf-multiplatform.nix)` seems to have references to native `coreutils`, `glibc` and whatnot, however, it seems to be an improvement.

Did you copy over the closure and see if nix behaves properly?

**Comment by eamsden** [comment]
I have not yet. Odd that the closure still pulls in native stuff. I'll try it tomorrow when I can get my pi back out

**Review by flokli — APPROVED** [review]

**Pull request merged**